libqtile/widget/redshift.py

```diff
@@ -0,0 +1,333 @@
+        """
+        if not self._lines:
+            return
+
+        obj = self._lines[self._line_index]
+
+        self.update(self.format_object(obj))
+
+    @expose_command
+    def enable_redshift(self):
+        """
+        Enable redshift with the temp, gamma and brightness params.
+        """
+        gamma_val = GammaGroup(self.gamma_red, self.gamma_green, self.gamma_blue)
+        try:
+            self._get_rs_driver().enable(self.temperature, gamma_val, self.brightness)
+        except Exception as e:
+            raise RuntimeError(f"redshift: could not enable redshift: {e}")
+
+    @expose_command
+    def click(self):
+        """
+        Either enable or disable the widget ONLY if we are currently on the
+        first line ie. the enabled/disabled text.
+        """
+        if self._line_index != 0:
+            return
+        if self.is_enabled:
+            self.reset_redshift()
+        else:
+            self.enable_redshift()
+        self.is_enabled = not self.is_enabled
+        self._set_text()
+        self.bar.draw()
+
+    @expose_command
+    def reset_redshift(self):
+        """
+        Make a call to redshift to reset everything.
+        """
+        try:
+            self._get_rs_driver().reset()
+        except Exception as e:
+            raise RuntimeError(f"redshift: could not reset redshift: {e}")
+
+    def _scroll(self, step):
+        """
+        Scroll up/down dictated by 'step' the items
+        and then display that item.
+        """
+        # Make sure that the lines var is set, maybe
+        # some of the values have been updated?
+        self._set_lines()
+        self._line_index = (self._line_index + step) % len(self._lines)
+        self.show_line()
+
+    def _set_lines(self):
+        """
+        Set the list of formatted text items to scroll through.
+        """
+        first_text = ""
+        if self.is_enabled:
+            first_text = self.enabled_txt
+        else:
+            first_text = self.disabled_txt
+
+        first_text = self._format_first_text(str(first_text))
+
+        self._lines = [
+            first_text,
+            self._format_brightness_text(),
+            self._format_temp_text(),
+            self._format_gamma_text(
+                GammaGroup(self.gamma_red, self.gamma_green, self.gamma_blue)
+            ),
+        ]
+
+    def _set_text(self):
+        """
+        Set text (not update) to the current index of lines.
+        """
+        if self._line_index == 0:
+            if self.is_enabled:
+                self.text = self._format_first_text(str(self.enabled_txt))
+            else:
+                self.text = self._format_first_text(str(self.disabled_txt))
+        else:
+            self.text = self.format_object(self._lines[self._line_index])
+
+    def _format_temp_text(self) -> str:
+        """
+        Format the temperature text.
+        """
+        return str(self.temperature_fmt).format(temperature=self.temperature)
```

Comment:
Does this need to be wrapped in `str`?

Reply:
So I could get away without forcing it to be a string( we know it is a string) but I think python doesn't because we're declaring it in the `defaults` rather than through `self.` so I have a bunch of LSP errors: 
```
Diagnostics:
1. "format" is not a known attribute of "None" [reportOptionalMemberAccess]
2. Cannot access attribute "format" for class "partial[Unknown]"
     Attribute "format" is unknown [reportAttributeAccessIssue]

```

This doesn't stop the code from functioning but I thought it would be better to silence it with `str()`

Reply:
@elParaguayo Am I ok to resolve this? The rest of the open issues are based on this

Reply:
I don't like the current code. Feels unnecessary but I've not had time to look at a solution.

Reply:
Fair enough, don't worry I'll just fix the forcing of `str` then

Reply:
Ok no worries

Reply:
@elParaguayo If you let me know your reasons why I can take a look myself

Reply:
I don't like forcing the type just to make the linter happy.

Would it work by adding `temperature_format: str` after the class definition?

Reply:
Yh you're right, let me give that a shot. What are your feelings on something like `assert isinstance(x, str)`?

Reply:
We've used those elsewhere to keep mypy happy so we could do it here but I still find that ugly. I'd prefer to see if we can set the type for that attribute separately and only use asset if the alternatives don't work.

Reply:
Accidentally deleted my previous comment but I have declared the defaults like: `temperature_format: str` and that seems to work